### ISSUE TITLE
GDScript: Design goals

### ISSUE THREAD

**Issue by vnen** [godotengine/godot#18698]
There are a lot of requests for changing stuff in GDScript, be it to extend functionality, make some syntax sugar, or add their favorite language feature. While it's good to know how people are using the language in order to improve it, deciding cases ad-hoc is prone to lead to inconsistency. It'll become more and more like a Frankenstein monster.

My proposal here is to create some master **design goals** for GDScript in the [roadmap repository](https://github.com/godotengine/godot-roadmap) and decisions should be based on such document. That way there's a clear set of criteria to accept or reject proposals made by users.

This ticket is to ask about ideas on how to make such document, and what goals we should have in mind. **This is not a place to ask for your wanted feature**, this is more of "meta design", showing which way we should go in general, and not specific things that should go in and out of GDScript.

For instance, there are some decisions to be made:
* GDScript should be kept as small as possible vs. embrace a large library of functions.
* Whether it should have powerful operators that acts on different types, or it should be limited to mathematical concepts.
* Add more Python-like statements, or go to a simpler route and add only what's hard/impossible to work around.
* More functional vs. more procedural approach.
* Only one way to do something vs. more shorthand constructs (AKA syntax sugar).
* Whether to embrace a variety of coding styles and paradigms (being more welcome for experience from other languages) or define a more strict GDScript-way of doing things (being easier to collaborate with different coders).

And probably some other stuff that I'm forgetting.

The idea here is not to decide which features will go in (though that's the endpoint) but to make up which criteria we should use to decide the features.

----
These are some ideas that I gathered when I was writing a proposal for discussion. I'll leave them here as it is what I now have in mind:

1. **GDScript should be expressive but not bloated.**
  * E.g. we probably don’t need a list comprehension syntax, but there should be an easy way to do it, in particular for making nested Arrays.
  * In fact, this relates directly to the Godot API. The more things available in the API, the less GDScript needs to do (also, it’ll be available for other languages as well).
2. **GDScript should try to be object-oriented, while allowing some functional paradigms.**
  * This means that inheritance, polymorphism, encapsulation principles should be applied.
  * The language should provide first-class functions and implementation of common higher-order functions (such as map, filter, and reduce).
  * Re-use of functionality is also expected. So it needs constructs outside of inheritance to share code (such as Traits).
3. **GDScript should be simple to write, hard to do wrong.**
  * Type specifiers are optional, but types should be strong.
  * The compiler must try to catch as many potential errors as possible, showing error and warning messages (in debug mode) for possible programming mistakes.
  * The compiler should make code optimizations based on inferred semantics, even when the programmer is being verbose. E.g. treating a local variable as constant if it’s never changed.
4. **GDScript implementation should be simple.**
  * So more people can understand and contribute. We don’t really need a super optimized language, though we do want to have optimizations.
  * Syntax should be simple and fit an LL(k) grammar with a small ‘k’ value. I'm managing to assume LL(1) for the new parser, and so far it works. So I'm avoiding any syntax that increases lookahead (as this makes things harder to parse).
  * Reduce the number of builtin functions to a minimum (use the engine API for more things).

**Comment by NathanLovato** [comment]
As a tutor the fact that GDscript is focused on gameplay programming at the moment is a great advantage. It does the job well yet it doesn't feel bloated and provided you're careful in how you implement things the code stays pretty easy to read.

I'd recommend to bring Lars Kokemohr in the loop: he's got a lot of gamedev experience, including AAA, thinks seriously about code architecture and languages, and he's got experience teaching Godot and Unity at university level so he's got great insights to offer.

**Comment by mhilbrunner** [comment]
>GDScript should be kept as small as possible vs. embrace a large library of functions.

*As small as possible* is exactly the crux of the matter, because *as possible* implies *as possible while still fullfilling its purpose*.

Also, these are different concerns: *a large library of functions* does not necessarily imply a complex language. And this is exactly where I would try to draw the line:

- Try to keep GDScript, the language, as small as possible
GDScript is hard to contribute to (and maintain), like the 3D renderer. Thus it scales badly with number of contributors. The larger and more complicated it becomes, the harder it is for others to improve upon it.
Favor performance and simplicity (and ease of use and learning) over features. We **can** add nice-to-haves, but should be conscious about the cost in terms of complexity.

- Favor adding methods to the standard library over additions to GDScript
A large standard library scales way better than new operators or statements for GDScript.
This is why I was in favor of adding e.g. new String utility methods in the past:
In contrast to GDScript, these have no additional cost for us when making other changes; if I now add something somewhere else, I probably don't need to think about the implications of those String methods. But when we add things to GDScript, that influences every past and future contribution to GDScript, most of the time.

**TL;DR:**

- Try to keep GDScript itself small, fast, easy to learn and maintain
- Aim for a comprehensive standard library instead, Python *batteries included* style

**Comment by mhilbrunner** [comment]
>Add more Python-like statements, or go to a simpler route and add only what's hard/impossible to work around.
>Only one way to do something vs. more shorthand constructs (AKA syntax sugar).

There should only be one (exposed and documented) way to do a thing, but that way should be as easy as possible.

>Whether to embrace a variety of coding styles and paradigms (being more welcome for experience from other languages) or define a more strict GDScript-way of doing things (being easier to collaborate with different coders).

Keep one *Godot way* to do things, like there is a *Pythonic way* to do things.
Don't make other styles impossible unless it gains performance/simplicity/ease of use, but don't encourage them officially either.

**Comment by NathanLovato** [comment]
Agreed with @mhilbrunner here. Especially on the:

> Keep one Godot way to do things, like there is a Pythonic way to do things.

Regarding the library part it means focusing on improving, adding nodes and other built-in classes right? Doing so benefits other languages too so it benefits every developer.

**Comment by nobuyukinyuu** [comment]
Please make these things more consistent:

1.  The use of properties in lieu of getters and setters still leans heavily in favor of the latter.  Usage could be reflective of the expected amount of impact or side effects reading or changing a value has, but it's not always clear.  

2.  Constructor syntax is wildly inconsistent.  Some types have functions which return an instance of it's type as a special overload of it's type name.  Others use the `.new()` or `.instance()` syntax, types of which aren't clearly documented.  Yet others still use an initializer function with a `void` type on a recently created instance.  Arrays and dictionaries have their own unique syntax. 
 This all might be alleviated with function overload support, which may require better type hinting to implement, which brings me to...

3.  More consistent type hinting.  Yes, gdscript is duck typed.  But hinting was clearly favored in some places early on and it would be helpful for users to have explicit syntax for it everywhere vars can be declared, not just in exports.  The syntax which already exists is already compatible with most use cases, save for explicit instancing because of the inconsistent ctor syntax.  Maybe it's my inexperience with dynamic typed languages leading to my confusion.  Can type-hinted exports can be `Nil` until explicitly initialized? Regular script variables seem to rely on determining its type at declaration based on assignment, and implicit type conversion behavior isn't yet clear to me.  Getting back to the point, making this behavior more obvious will help make it easier – or possible, if it isn't already – to unify the syntax used to give a recently declared variable a value.

Hopefully neither of these points ventured too far into feature request territory.  Some things seemed necessary to implement to meet these design suggestions as a matter of course, but may not be necessarily reflective of the major contributors' design goals for the language...

**Comment by mhilbrunner** [comment]
@homer666 Have to agree. Those were nice changes for sure.

**Comment by Zylann** [comment]
I miss functional things a little, especially lambdas which were proposed 3 times now but for some reason never went merged... maybe not enough time to review / big updates were coming at the same time?

Also, I miss an `await`, or perhaps a better `yield`. When you deal with timed things such as requests or even dialogs, using callbacks all over the place gets messy really quick (since all vars need to be flattened as class members and you get more functions in the same scope).

**Comment by reduz** [comment]
I mostly agree with @vnen. This is more a practical than a theoretical thing:

GDScript codebase needs to stay small to be maintainable. The more features we add, the more unmaintainable it becomes. Features that you may use only a few times in your project that will save 2 or 3 lines of code, should not be included. 

If adding a cool feature that will be useful a few times only takes like, 1 or 2 lines of code.. this is fine, nothing against it. If adding it takes several dozen lines of code, then it will most likely be rejected.

Property access and the $ operator were features you use often that definitely improve quality of life. The upcoming optional typing is the same.. so even if they need a few dozen lines of code, they are worth it.

Lambdas may also be useful, but if it means hugely increasing GDScript size, then they should not be added. 

The best approach that needs to be taken here to implement features is to do it using the least amount of new code possible.

**Comment by mhilbrunner** [comment]
> I think it should track python as closely as possible.

Disagree, Python is a different language with way different goals.
Take the ideas that fit and sure, let GDScript be inspired by it, but don't cling to it too hard.

But yeah, lambdas/first class functions and variable amount of arguments for functions (variadic functions) are the two biggest things I miss, Zylann. :P Sorry, thats kind of "feature wishlist" territory again.

Again: Happy birthday, @reduz :)

**Comment by lorenzobeccaro** [comment]
> This is great in theory, but in practice it makes the codebase more difficult to maintain. I think the fact Python will soon be an optional language will be a solution for developers who really can't live without the more advanced syntax sugar.

Exactly what I meant. 
Also.. Happy birthday :)

**Comment by mwerezak** [comment]
> Keep one Godot way to do things, like there is a Pythonic way to do things.
Reminds me of something I was working with yesterday. When we want the size of an Array, do I use `array.size()` or `len(array)`. When I want to append to the end of an array, do I `array.append(item)` or `array.push_back(item)`.

Personally I prefer using `array.size()`. It's more consistent, just a method call like any other. While `len(array)` seems to be some kind of special language construct that AFAICT exists only to mimic Python.

And similarly with `array.push_back(item)`. There's already a `push_front()` method so it seems more consistent.

Of course these are very specific examples, but I just wanted to highlight something worth considering.

**Comment by vnen** [comment]
Thanks for all the responses.

First, let me clarify a few points that seem to stir confusion:

> There are 3 different syntax for constructors.

Not really. One are for primitive types, which are not "instances" per se (like `Color()`). One you create a new instance of an object (`Object.new()`). The other one is not a constructor, it is meant to get the scene out of a PackedScene resource ( `scene.instance()`, which is just a regular method from PackedScene).

> `Array.append()` and `Array.push_back()` do the same thing.

This is more about how you can use the array as queue, list, or stack using more conventional function names. But this should be another discussion since they are not GDScript constructs, they are part of the Godot API. There's an important difference there, because changing this API requires no change to GDScript itself.

> The choices aren't mutually exclusive.

My list was more about examples for discussion than strict decisions to make. Many of those can be viewed as a spectrum rather than a discrete boolean choice.

> I miss `X` from my experience with other languages.

That's quite a complicated topic and it relates to my last bullet from the OP. The thing is that while you might want to do it in the way you're used to, there might be another (more Godot-y) way to do the same thing. Hence my question: should we embrace many paradigms or solve this by documenting the translation between those and the Godot way? For instance, signals are very important in Godot but many people seem to forget they exist.

> About the `len()` function.

This is a GDScript function but it serves more to generalize getting the length of different types of collections (for instance, `Array.size()` vs. `String.length()`). Though this might only be solving a deficiency in the API itself, which can be discussed. This might apply to a few other built-in function as well.

----

From what I can gather from this discussion (and the previous history) so far:

* People are not so fond of syntax sugar, but a few constructs are really wanted.
* Preference seems to be a small but expressive language.
* Less ways to do the same stuff are better and that way should be simple.
* Add stuff to the main Godot API instead of a new constructs/functions in GDScript. I just want to note that adding to the API has its own criteria.
* There are still sorely missed features in GDScript that make it not expressive enough (variadic functions/splat operator, first-class functions/lambdas, null coalescing operator, and maybe a few others).

Also, while I do like embracing the functional paradigm, if we add first-class functions we'll probably need more functional stuff as well (such as `map()` and `reduce()` for collections, currying/partial application). And that will affect that typing system that's on the way (type hints might include not only that it is a "function" but the signature as well). This is something to consider.

**Comment by triplefox** [comment]
I have a recommendation and that is to look for some metric of improvement(e.g. ratio of LOC added in GdScript to LOC saved in example scripts). This is a technique I have read that Niklaus Wirth used in his languages, with the main metric in his case being the speed with which the compiler could compile itself. This led to tradeoffs between source code size and algorithmic optimizations, with simpler, cleaner  algorithms often winning out when put to the test.

Doing this will help in rejecting bloat going forward, while still allowing for some evolution. Features that eliminate whole classes of errors in user code would override a rule like this(maybe there is a way to capture potential error rate in a number?)

**Comment by mhilbrunner** [comment]
IMHO LOC is a bad metric. It only measures volume, not quality. It doesn't capture whether the example code improves, or becomes more readable and maintainable.

Shorter code isn't necessarily better. See arguments about better variables names.

**Comment by vnen** [comment]
I assume LOC was only an example of metric, not an actual proposal. The idea of a metric is interesting, but I can't think of anything that would fit nicely into "the Godot way" of doing things.

**Comment by razvanc-r** [comment]
This is probably controversial, but in my opinion I think that moving towards reactive (not necessarily functional, although it kinda' comes with the territory) is what needs to happen. For multiple reasons, but most of all it's just natural to work with when dealing with time and multimedia interactions. It's the response to contemporary needs and most languages are moving in this direction, mostly through libraries, but there are a few that have built-in functionality (like [dart](https://www.dartlang.org/tutorials/language/streams)). For generic info on reactive programming and why it's so useful [check this](https://gist.github.com/staltz/868e7e9bc2a7b8c1f754).

This imperative type of programming is suited very well for *input -> crunch -> output* type of applications, not interactive stuff.

Think of the mouse, instead of getting one value at a time (through connecting to the event) and being forced to keep track of the time yourself and do maths with time, control flow etc. is very unnatural. Instead, with composable streams where you have "atomic" functions that operate on incoming data as data is fed into the pipe is far more natural, it's exactly what you'd expect to do, instead of keeping track of each individual event and having to think about all the interactions with future/past events, keeping track of them individually, a stream that takes say for example the events in the last 0.5 sec and gives them to you to process or combine multiple streams or just give you double taps as an event are much more natural.

So I do agree with everyone before me about keeping it simple and maintainable, I just don't think that sticking with the imperative style is suited for interactive applications. It can be done sure but with a lot of boilerplate and it's very easy to get wrong the order in which you process stuff as things get passed around all over the place randomly really, if you think about the user interaction with the game.

**Comment by voithos** [comment]
There has been a lot of discussion so far, and from what I gather it seems that we've been able to narrow down on some design goals for GDScript. Is there more to discuss, or is this ready to be concretized in the form of some docs in the roadmap repo, as originally mentioned?

For what it's worth, I agree with a lot of what has already been mentioned, including the summary in https://github.com/godotengine/godot/issues/18698#issuecomment-388073116. I would suggest, though, adding "_consistency_" (both in APIs as well as in GDScript itself) as a goal in addition to the ones you already listed.

#### Rationale:
Consistency in syntax and naming decreases cognitive load and allows developers to build more of an intuition around how the system works. In some cases, it may be justifiable to break consistency in order to (for example) highlight a truly unique case or make slow / "dangerous" operations easier to detect - but in the general case, I think we should strive for consistency.

For example, as mentioned, the 3 different syntaxes for "constructors" apply to different cases / types of entities, but we should consider whether the differences are important enough to justify different syntax. As a developer using Godot, do I really need to care that `Color` is primitive and `Object` isn't, such that I have to do `Color()` vs `Object.new()`? If there is a good reason for the developer to care, then perhaps the different syntax is justified - but otherwise, they would ideally be made more consistent.

**Comment by TheYokai** [comment]
In terms of future elements that godot could have, a short hand for `funcref()` might be a useful way to provide some kind of functional support without having to support lambdas immediately. I'm thinking of something similar to the way that `$` works, maybe using an `->` syntax. For instance:

```
self->gotoscene
```

Would basically be the equivalent of writing something like....

```
funcref( self, "gotoscene" )
```

Idk how others would feel about this, but it could be kind of useful...

**Comment by Zylann** [comment]
I don't think it's worth adding a special operator just to shorten `funcref`, the latter is already quite short to write. I'd prefer actual lambda support. At best, for a shortcut, the syntax `object.method` without `()` could be a better step towards first-class functions.

**Comment by aaronfranke** [comment]
What do the discussions in this thread mean for issues such as #23101? Would such complex things ever be added if they are only useful in some situations for some users?

**Comment by Zylann** [comment]
@aaronfranke this comment is quite relevant regarding typed GDScript https://github.com/godotengine/godot/issues/23101#issuecomment-430833465

**Comment by vnen** [comment]
Adding this to the OP as well for visibility, but these are my current ideas:

1. **GDScript should be expressive but not bloated.**
  * E.g. we probably don’t need a list comprehension syntax, but there should be an easy way to do it, in particular for making nested Arrays.
  * In fact, this relates directly to the Godot API. The more things available in the API, the less GDScript needs to do (also, it’ll be available for other languages as well).
2. **GDScript should try to be object-oriented, while allowing some functional paradigms.**
  * This means that inheritance, polymorphism, encapsulation principles should be applied.
  * The language should provide first-class functions and implementation of common higher-order functions (such as map, filter, and reduce).
  * Re-use of functionality is also expected. So it needs constructs outside of inheritance to share code (such as Traits).
3. **GDScript should be simple to write, hard to do wrong.**
  * Type specifiers are optional, but types should be strong.
  * The compiler must try to catch as many potential errors as possible, showing error and warning messages (in debug mode) for possible programming mistakes.
  * The compiler should make code optimizations based on inferred semantics, even when the programmer is being verbose. E.g. treating a local variable as constant if it’s never changed.
4. **GDScript implementation should be simple.**
  * So more people can understand and contribute. We don’t really need a super optimized language, though we do want to have optimizations.
  * Syntax should be simple and fit an LL(k) grammar with a small ‘k’ value. I'm managing to assume LL(1) for the new parser, and so far it works. So I'm avoiding any syntax that increases lookahead (as this makes things harder to parse).
  * Reduce the number of builtin functions to a minimum (use the engine API for more things).

**Comment by aaronfranke** [comment]
> @vnen E.g. treating a local variable as constant if it’s never changed.

This reminds me, it would be really nice to have local `const` fields. They would be forcefully unchangeable like class-level constants are, but limited to the scope they're declared in.

I agree with all your points, but I'm a bit confused as to what "LL(k)" means. Does LL(1) mean that any given token's behavior is not changed by the following token? That sounds good to me.

**Comment by NathanLovato** [comment]
I second @aaronfranke on local constants. In some longer files we've worked on lately, we had quite a few cases where we needed just that (specific tween-based animations with different durations and delays, among other things)

**Comment by Calinou** [comment]
@aaronfranke There's a proposal about immutable variables here: https://github.com/godotengine/godot-proposals/issues/820

**Comment by vnen** [comment]
@aaronfranke 
> This reminds me, it would be really nice to have local `const` fields. They would be forcefully unchangeable like class-level constants are, but limited to the scope they're declared in.

I'm already doing this in the GDScript rewrite.

> [...] I'm a bit confused as to what "LL(k)" means. Does LL(1) mean that any given token's behavior is not changed by the following token? That sounds good to me.

That's indeed a quite technical term. One `L` means that it parses left-to-right, the other means "lookahead", and `k` is the amount of tokens of lookahead need to decide the grammar production (e.g. whether it's a variable or function declaration or something else). LL(1) means the parser needs to look at the next 1 token to decide which function to run. A larger number means it has to buffer more tokens and have more possibilities to decide the next function.

Note: It is possible to have an LL(0) language, which is super simple to parse, but it's very difficult to make it look natural and useful.

**Comment by hilfazer** [comment]
Be newbie friendly.

It already is and should stay that way while becoming even friendlier.
So for example i'd like it to avoid confusing syntax like { = }
```
var s = "foo"
var d = { s = 5 }
var f = { s : 5 }
```
Newbie can be unaware of optionally skippable quotes. Or s/he could not notice there's `=` and not `:` (or vice versa).

Or here - newbie can think $ is fetching a node named `bar`:
```
	var foo = "bar"
	$foo
```

**Comment by xix-xeaon** [comment]
I'd like to point out an other area of focus that is rarely actually talked about which is **productivity**. Having a hammer which is easy to learn how to use, or easy to construct/repair for the manufacturer is all well and good, but the hammer exists to _be used to achieve some result_. A hammer which is more difficult to learn, or harder to make, is absolutely worth it **if** it produces better results.

The point of scripting languages is not really to be easy to learn but instead to be _easy to work with_ in order to get high productivity. (This is also what _makes_ them easy to learn.) You want to be able to quickly write only a small amount of code and get your idea working. Compiled languages instead focus on runtime performance.

Python is the epitome of productive languages and GDScript is inspired by it because it shares the same strong focus on productivity, with the desire to be able to quickly write game logic and mechanics. It is important to remember this when considering what features to add.

It's true that maintaining all of the features of Python in GDScript would certainly be too much work, and some things wouldn't even fit at all because of performance since game code often needs to run every single frame. It is, however, not true that Python is difficult to learn - despite all those features. This is because you don't have to use features you don't know.

**With this in mind**
I certainly agree on having a large standard library, since it increases productivity. Also, as has been pointed out, it doesn't make the language itself more complex so it doesn't add much to maintenance. The requirement for inclusion should be very low - even if it only helps a few only a little.

For instance, Array.extend and Dictionary.update. Making those functions yourself is only 4 lines of code each and most people don't need them most of the time. But including them costs nothing and it increases productivity. I'm guessing there'd also be quite a difference in performance. An other example is a lack of a simple builtin function for listing a directory. Python is full of functions like these and GDScript should be too.

On the subject of list comprehension (with dictionaries too); I know some people _really_ don't like them, and that's fine - you don't have to use them - but they're very productive if you're comfortable with them. Also, generators/iterators from yield in functions and array-slicing syntax. And sets. And itertools etc. Often you're only working with individual things but as soon as you have a bunch of something all of these functions and features become very useful. It's true that they can be hard to wrap your head around for beginners, and I don't know how much those features would affect the parser, but they're just so productivity once you learn them.

Keyword arguments (optional unordered arguments with defaults) is an other example of an extremely productive feature, even for beginners. It almost doesn't matter what the implementation cost is, because it makes it so easy to create objects and call functions with only the arguments you want. With it you can have all the .new() of nodes etc accept the three parameters you want to change right away instead of having them on three extra lines afterwards.

Then there are things like custom datatypes (for overloading operators). Beginners won't use them, that's fine. But in some situations it makes all the other code so much easier to work with, greatly increasing productivity. This feature has performance implications for other code, even when not used, but I think it can be implemented separated from the normal object.

Again these are just examples, but I think **productivity** should absolutely be a core part of the design goals. GDScript exists to get more work done faster.

**Comment by vnen** [comment]
> Newbie can be unaware of optionally skippable quotes. Or s/he could not notice there's `=` and not `:` (or vice versa).

Honestly a newbie will be unaware of many things, because they are a newbie. Such person should be reading documentation and tutorials to learn. The GDScript basics page is not that long (sure, it has a lot to soak in, but such mistakes can be avoided and understood by reading the relevant section again). That's not an excuse to bloat GDScript but we should not strip it and remove features from the pros to write optimal code just because a newbie might stumble on it.

BTW the `{ a = 1 }`  syntax is based on the Lua tables, which is supposed to be a basic language. It helps writing dictionaries when every key is a string.

> I certainly agree on having a large standard library, since it increases productivity. Also, as has been pointed out, it doesn't make the language itself more complex so it doesn't add much to maintenance. The requirement for inclusion should be very low - even if it only helps a few only a little.

Note that adding things to Godot should follow the proposal workflow. And if something is not used often it won't make the cut. The point is that those things do not need to be provided out of the box from Godot but instead provided as plugins that you can download from the Asset Library. So only those who need it will get it.

> For instance, Array.extend and Dictionary.update. Making those functions yourself is only 4 lines of code each and most people don't need them most of the time. But including them costs nothing and it increases productivity. I'm guessing there'd also be quite a difference in performance. An other example is a lack of a simple builtin function for listing a directory. Python is full of functions like these and GDScript should be too.

Well, it costs almost nothing to add one thing. But then one by one this becomes a lot of things, and then it does make a difference. How often you need to list a directory in a game project? If you need to process the entries, iterating a Directory object won't be much different than iterating a list you got with a builtin function. Also is it recursive or not? Should it show hidden (dot) files? As you start adding arguments to customize that, the function isn't so simple anymore (and if you don't add those then it's less useful thus less used). It's not that simple to add stuff without thinking of use cases, that's why we introduced the proposals.

> Keyword arguments (optional unordered arguments with defaults) is an other example of an extremtely productive feature, even for beginners. It almost doesn't matter what the implementation cost is, because it makes it so easy to create objects and call functions with only the arguments you want.

Adding keyword arguments would make the release binary bigger and it would introduce a performance cost every time you use them. You know what's gonna happen? It will get a bad rap for the performance penalty and people will start recommending not using it, which renders the feature pointless.



> Again these are just examples, but I think **productivity** should absolutely be a core part of the design goals. GDScript exists to get more work done faster.

My point is that GDScript should be easy to write, which kind of implies that we want people to be productive with it. But it should not happen in detriment of other points. Even though performance is not the goal above it all, we should not make it slow for an edge-case feature.

And again: everybody can open a [proposal](https://github.com/godotengine/godot-proposals) for a particular feature they need. My ideal for this "design goals" is to be one parameter for the GDScript-related proposals to be considered (notice that I opened this issue way before the proposal system was in place). It's mostly guidelines that are a tad subjective. So most of the examples could actually be added, as long as they go through the workflow we set up. This is not a dismissal for the examples, just an explanation that it's not that simple to "just add everything that is marginally helpful" as that could have an impact in other areas even for people who don't use it.

**Comment by Calinou** [comment]
Closing, as GDScript 2.0 is now feature-complete. Further changes should be discussed on the [Godot proposals](https://github.com/godotengine/godot-proposals) repository.
